enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,421 @@
+---
+title: expose-compression-variables-in-haproxy
+authors:
+- "@candita"
+
+reviewers:
+- "@frobware"
+- "@knobunc"
+- "@Miciah"
+- "@miheer"
+- "@rfredette"
+- "@deads2k"
+
+approvers:
+- "@Miciah"
+- "@frobware"
+
+creation-date: 2021-08-27
+
+last-updated: 2021-08-27
+
+status: implementable
+
+see-also:
+
+replaces:
+
+superseded-by:
+
+---
+
+# Expose Compression Variables in HAProxy
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposal extends the IngressController API to allow the cluster administrator
+to configure HTTP traffic compression via HAProxy variables.  HTTP traffic compression increases the
+performance of a website by reducing the size of data that is passed in HTTP transactions.  
+
+## Motivation
+
+Customers with large amounts of compressible routed traffic need the ability to gzip-compress
+their ingress workloads, as described in feature request [NE-542](https://issues.redhat.com/browse/NE-542).
+HAProxy in OpenShift 3.x exposed compression variables, so lack of access to the variables in OpenShift 4.x
+is a migration blocker for customers that were using them in OpenShift 3.x.
+
+The ability to configure compression was available in 3.x, but it applied only to the HAProxy
+`defaults` configuration section, and supported only the `gzip` compression algorithm.  This proposal
+reintroduces and then enhances the previous implementation in these ways:
+- allow the `frontend`, and `backend` sections in addition to the `defaults` section, which would provide for more
+  fine-grain configuration at the route level.
+- allow a choice of `gzip` or `raw-deflate` compression algorithms.
+
+### Goals
+
+Enable the configuration of HTTP compression global defaults via the `IngressControllerSpec`.
+
+Enable the configuration of HTTP compression for route backends/frontends via route annotations.
+
+Enable HTTP compression on selected MIME types, be they wildcard (e.g. text/*, application/*, etc.) or
+specific (e.g. text/html, text/plain, text/css, application/json, image/svg+xml, font/eot, etc.).  Not all MIME
+types benefit from compression, but there is a short list found [here](https://letstalkaboutwebperf.com/en/gzip-brotli-server-config/)
+that will be implemented as a part of this proposal.
+
+By exposing ROUTER_ENABLE_COMPRESSION, any HTTP request with the header "Accept-Encoding: gzip" or
```

Comment:
This will only apply to routes that the router terminates right?  So not pass through.

Reply:
Thanks, I clarified that.